src/Shared/PackageManagers/PyPIProjectManager.cs

```diff
@@ -432,6 +432,37 @@ public override async Task<IEnumerable<string>> EnumerateVersionsAsync(PackageUR
             return metadata;
         }
 
+        public override async Task<DateTime?> GetPublishedAtUtcAsync(PackageURL purl, bool useCache = true)
+        {
+            Check.NotNull(nameof(purl.Version), purl.Version);
+            HttpClient client = CreateHttpClient();
+
+            JsonDocument contentJSON = await GetJsonCache(client, $"{ENV_PYPI_ENDPOINT}/pypi/{purl.Name}/{purl.Version}/json");
+            JsonElement root = contentJSON.RootElement;
+
+            JsonElement.ArrayEnumerator? urlsArray = OssUtilities.GetJSONEnumerator(root.GetProperty("urls"));
+            DateTime? uploadTime = null;
+            if (urlsArray is not null)
+            {
+                foreach (JsonElement url in urlsArray.Value)
+                {
+                    string? urlStr = OssUtilities.GetJSONPropertyStringIfExists(url, "url");
+                    string? uploadTimeStr = OssUtilities.GetJSONPropertyStringIfExists(url, "upload_time");
+                    if (uploadTimeStr != null)
+                    {
+                        DateTime newUploadTime = DateTime.Parse(uploadTimeStr).ToUniversalTime();
+                        // Used to set the minimum upload time for all associated files for this version to get the publish time.
+                        if (uploadTime == null || uploadTime > newUploadTime)
+                        {
+                            uploadTime = newUploadTime;
+                        }
+                    }
+                }
+            }
```

Comment:
Should we also extract the above logic into a helper method (like what you did for NPM), to avoid duplication?

Reply:
I looked at it, the parsing logic to get the upload time involves parsing a json array which has other metadata apart from the upload time. If we use a helper method to move the logic to extract upload time from json array, that logic is duplicated in the GetMetadata method as we need the array parsing to extract other metadata. There is only 3-4 lines of code to get the minimum upload time that can be extracted to helper method. I thought it should be ok to leave it as is.

Reply:
OK, that makes sense. Thanks for looking into it.